docs/sql/02-type-system.md

```diff
@@ -0,0 +1,175 @@
+
+Every input value is mapped to an SQL type first. If there is no appropriate Java-to-SQL mapping, then the value is
+interpreted as the `OBJECT` type. Then the value is converted to a Java type mapped to the SQL type. Internally the
+engine operates only on Java types defined in SQL-to-Java mapping, which simplifies the implementation significantly.
+
+For example, an input value of the type `java.math.BigInteger` is mapped to `DECIMAL` SQL type. But since `DECIMAL`
+type is mapped to `java.math.BigDecimal`, the input value is converted from `java.math.BigInteger` to `java.math.BigDecimal`
+on first access.
+
+### 2.1 SQL to Java Mapping
+
+Table 2 establishes a strict one-to-one mapping between SQL and Java types.
+
+*Table 2: SQL-to-Java mapping*
+
+| SQL Type          | Java Type                  |
+|-------------------|----------------------------|
+| `NULL`            | `java.lang.Void`           |
+| `VARCHAR`         | `java.lang.String`         |
+| `BOOLEAN`         | `java.lang.Boolean`        |
+| `TINYINT`         | `java.lang.Byte`           |
+| `SMALLINT`        | `java.lang.Short`          |
+| `INTEGER`         | `java.lang.Integer`        |
+| `BIGINT`          | `java.lang.Long`           |
+| `DECIMAL`         | `java.math.BigDecimal`     |
+| `REAL`            | `java.lang.Float`          |
+| `DOUBLE`          | `java.lang.Double`         |
+| `DATE`            | `java.time.LocalDate`      |
+| `TIME`            | `java.time.LocalTime`      |
+| `TIMESTAMP`       | `java.time.LocalDateTime`  |
+| `TIMESTAMP W/ TZ` | `java.time.OffsetDateTime` |
+| `OBJECT`          | `java.lang.Object`         |
+
+`TIMESTAMP WITH TIME ZONE` is mapped to the `java.time.OffsetDateTime` class because ANSI SQL requires only zone
+displacement, so full zone information from the `java.time.ZonedDateTime` class is not needed.
+
+### 2.2 Java to SQL Mapping
+
+Table 3 establishes a many-to-one mapping between Java and SQL types.
+
+*Table 3: Java-to-SQL mapping*
+
+| Java Type                  | SQL Type          |
+|----------------------------|-------------------|
+| `java.lang.Void`           | `NULL`            |
+| `java.lang.String`         | `VARCHAR`         |
+| `java.lang.Character`      | `VARCHAR`         |
+| `java.lang.Boolean`        | `BOOLEAN`         |
+| `java.lang.Byte`           | `TINYINT`         |
+| `java.lang.Short`          | `SMALLINT`        |
+| `java.lang.Integer`        | `INTEGER`         |
+| `java.lang.Long`           | `BIGINT`          |
+| `java.math.BigInteger`     | `DECIMAL`         |
+| `java.math.BigDecimal`     | `DECIMAL`         |
+| `java.lang.Float`          | `REAL`            |
+| `java.lang.Double`         | `DOUBLE`          |
+| `java.time.LocalDate`      | `DATE`            |
+| `java.time.LocalTime`      | `TIME`            |
+| `java.time.LocalDateTime`  | `TIMESTAMP`       |
+| `java.util.Calendar`       | `TIMESTAMP W/ TZ` |
+| `java.util.Date`           | `TIMESTAMP W/ TZ` |
```

Comment:
the same as above: why use shortcut? I believe `TIMESTAMP W/ TZ` is not valid SQL